### ISSUE TITLE
aslanbek - DoS of `withdrawAllUnlockedToken` for any user via ~250 1-wei locks

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-05-gamma-staking-judging#156]
aslanbek

medium

# DoS of `withdrawAllUnlockedToken` for any user via ~250 1-wei locks

## Summary

`withdrawAllUnlockedToken` will not fit into 30M block gas limit for the caller once they have ~250 active locks.

## Vulnerability Detail

`withdrawAllUnlockedToken` iterates through every lock of the caller, and claims the ones that are `claimable` (which have previously been queued for withdrawal AND which unlockTime is reached).

However, it is not done just once for each lock, but the same operation is repeated once for every 10 locks (which does not change the state and just wastes gas), which makes the gas cost of this function grow quadratically with the number of locks.

As it is possible for any account to create locks for any other account, an attacker may stake ~250 stakes of 1 wei for the longest duration to DoS `withdrawAllUnlockedToken` for the victim for that duration.

Alternatively, if there's an account that wants to stake GAMMA daily for the longest duration, it will run into that problem without external influence.

## Impact

`withdrawAllUnlockedToken` reverts with OOG for the victim. In practice, if the victim uses the frontend, it would offer them to sign a transaction that would spend `block gas limit` of gas and revert.

## Code Snippet

https://github.com/sherlock-audit/2024-05-gamma-staking/blob/main/StakingV2/src/Lock.sol#L415-L426

## Tool used

Manual Review

## Recommendation

Do not allow accounts to stake for others, and refactor `withdrawAllUnlockedToken`, so it iterates through each lock only once:

```diff
-       uint256 page;
-       uint256 limit;
-       while (limit < lockCount) {
            LockedBalance[] memory lockedBals = locklist.getLocks(
                msg.sender,
                page,
                lockCount
            );
            for (uint256 i = 0; i < lockedBals.length; i++) {
                if (
                    lockedBals[i].unlockTime != 0 &&
                    lockedBals[i].unlockTime < block.timestamp
                ) {
                    totalUnlocked += lockedBals[i].amount;
                    locklist.removeFromList(msg.sender, lockedBals[i].lockId);
                }
            }

-           limit += 10;
-           page++;
-       }
```

P.S. the number 250 is an estimation that comes from `forge test --gas-report`, which shows the following gas cost for the number of 1-wei locks:
| locks | gas |
|-----|-------------|
| 240 | 28284685 |
| 230 | 25005480 | 
| 220 | 22049040 |
| 210 | 19390757 |
| 200 | 17007123 |
| 100 | 3853937 |

And for `locks = 250` yields `gas = 9079256848711094572`.

Duplicate of #214
